File: 0376-wiggle-subsequence/0376-wiggle-subsequence.cpp
class Solution {
public:
    int wiggleMaxLength(vector<int>& nums) {
        if (nums.size() == 1) {
            return 1;
        }
        
        int count = 1;
        int diff = 0;
        
        for (int i = 1; i < nums.size(); i++) {
            int currDiff = nums[i] - nums[i - 1];
            if ((currDiff > 0 && diff <= 0) || (currDiff < 0 && diff >= 0)) {
                count++;
                diff = currDiff;
            }
            
        }

        return count;
    }
};

File: 0376-wiggle-subsequence/0376-wiggle-subsequence.py
class Solution:
    def wiggleMaxLength(self, nums: List[int]) -> int:
        n = len(nums)

        if n <= 1:
            return n
        
        prevDiff = 0
        currDiff = 0
        count = 1

        for i in range(n - 1):
            currDiff = nums[i + 1] - nums[i]

            if currDiff > 0 and prevDiff <= 0 or currDiff < 0 and prevDiff >= 0:
                count += 1
                prevDiff = currDiff

        return count